reframe/core/schedulers/sge.py

```diff
@@ -0,0 +1,162 @@
+        )
+
+        # Options starting with `-` are emitted in separate lines
+        rem_opts = []
+        verb_opts = []
+        for opt in (*job.sched_access, *job.options, *job.cli_options):
+            if opt.startswith('-'):
+                rem_opts.append(opt)
+            elif opt.startswith('#'):
+                verb_opts.append(opt)
+            else:
+                select_opt += ':' + opt
+
+        return [self._format_option(select_opt),
+                *(self._format_option(opt) for opt in rem_opts),
+                *verb_opts]
+
+    def emit_preamble(self, job):
+        preamble = [
+            self._format_option('-N "%s"' % job.name),
+            self._format_option('-o %s' % job.stdout),
+            self._format_option('-e %s' % job.stderr),
+            self._format_option('-wd %s' % job.workdir),
+        ]
+
+        if job.time_limit is not None:
+            h, m, s = seconds_to_hms(job.time_limit)
+            preamble.append(
+                self._format_option('-l h_rt=%d:%d:%d' % (h, m, s)))
+
+        preamble += self._emit_lselect_option(job)
+
+        return preamble
+
+    def submit(self, job):
+        # `-o` and `-e` options are only recognized in command line by the PBS,
+        # SGE, and Slurm wrappers.
+        cmd = f'qsub -o {job.stdout} -e {job.stderr} {job.script_filename}'
+        completed = _run_strict(cmd, timeout=self._submit_timeout)
+        jobid_match = re.search(r'^Your job (?P<jobid>\S+)', completed.stdout)
+        if not jobid_match:
+            raise JobSchedulerError('could not retrieve the job id '
+                                    'of the submitted job')
+
+        job._jobid = jobid_match.group('jobid')
+        job._submit_time = time.time()
+
+    def poll(self, *jobs):
+        if jobs:
+            # Filter out non-jobs
+            jobs = [job for job in jobs if job is not None]
+
+        if not jobs:
+            return
+
+        user = osext.osuser()
+        completed = osext.run_command(
+            f'qstat -xml -u {user}'
+        )
+
+        if completed.returncode != 0:
+            raise JobSchedulerError(
+                f'qstat failed with exit code {completed.returncode} '
+                f'(standard error follows):\n{completed.stderr}'
+            )
+
+        root = ET.fromstring(completed.stdout)
+
+        # Store information for each job separately
+        jobinfo = {}
+        for queue_info in root:
+
+            # Reads the XML and prints jobs with status belonging to user.
+            if queue_info is None:
+                raise JobSchedulerError('Decomposition error!\n')
+
+            for job_list in queue_info:
+                if job_list.find("JB_owner").text != user:
+                    # Not a job of this user.
+                    continue
+
+                job_number = job_list.find("JB_job_number").text
+
+                if job_number not in [job.jobid for job in jobs]:
```

Comment:
This check is not very efficient. I would suggest keeping the jobs submitted by ReFrame in a set that you populate as you `submit()` them and here you simply check `if job_number in self._submitted_jobs`

Reply:
The only thing I don't like about this is that in theory now `poll` can work with jobs submitted by a different instance of a sge scheduler (I think). Not that we use this like that, but this is how it is in other schedulers as well. To avoid going through the list every time maybe a compromise is to make a set in the beginning of the `poll`.

Reply:
I updated my proposal. You don't need to keep track of the submitted jobs anymore, but simply index the jobs passed to `poll()`, because otherwise the complexity of polling becomes O(N*P), where N is the total jobs in the queue and `P` the number of jobs to poll.